packages/vite/src/node/plugins/html.ts

```diff
@@ -959,7 +959,13 @@ export function htmlEnvHook(config: ResolvedConfig): IndexHtmlTransformHook {
   for (const key in config.define) {
     if (key.startsWith(`import.meta.env.`)) {
       const val = config.define[key]
-      env[key.slice(16)] = typeof val === 'string' ? val : JSON.stringify(val)
+      if (typeof val === 'string') {
+        try {
+          env[key.slice(16)] = JSON.parse(val)
+        } catch {} // ignore non-JSON.parse-able values
+      } else {
+        env[key.slice(16)] = JSON.stringify(val)
+      }
```

Comment:
Or maybe we should just support strings?
```suggestion
      if (typeof val === 'string') {
        try {
          env[key.slice(16)] = JSON.parse(val)
        } catch {} // ignore non-JSON.parse-able values
      }
```

Reply:
Yes, I think it is better we only support stringified object/strings. It isn't intuitive from a HTML-only perspective, but IMO it is better to be consistent between JS and HTML here.
I wonder if we really need to support `define` for HTML 🤔

Reply:
Non JSON.stringified strings break the code already [here](https://github.com/vitejs/vite/blob/main/packages/vite/src/client/env.ts#L17).

So I think it is fair to throw an error for not stringified strings

Reply:
I think this shows that supporting `define` in HTML isn't being used as intended though. We wanted to [only add support for replacing env variables](https://github.com/vitejs/vite/issues/3105#issuecomment-1441947641). And you are going through a `define` to override a non-existent env variable to be able to use this feature. I think it is better for you to use an inline plugin and [`transformIndexHtml`](https://vitejs.dev/guide/api-plugin.html#transformindexhtml).

Reply:
> Supporting `define` in HTML feels like formalizing that define for `import.meta.env.VITE_XXX` is not a replacement that happens before the regular env replacement, but a way to overwrite env variables in the config.

The latter is actually what we're doing at https://github.com/vitejs/vite/blob/228ee35a38ccd366fc9fe4b25f2344e699f213b9/packages/vite/src/node/plugins/define.ts#L41-L46

So I'd say we could support `define` in HTML, and sapphi's change in this PR looks fine to me.

Reply:
I think overloading `define` to extend `import.meta.env.*` is fine since it's not a common thing to do, but happy to discuss more about it if it's confusing. The HTML case is indeed a bit tricky but I think a fix like this could be good enough for now.

Reply:
I'm leaning towards applying this suggestion (https://github.com/vitejs/vite/pull/13425#discussion_r1217804441).
We document the behavior as ["Any properties in `import.meta.env` can be used in HTML files with a special `%ENV_NAME%` syntax"](https://vitejs.dev/guide/env-and-mode.html#html-env-replacement:~:text=Any%20properties%20in%20import.meta.env%20can%20be%20used%20in%20HTML%20files%20with%20a%20special%20%25ENV_NAME%25%20syntax%3A). I understood it to mean replacing `%ENV_NAME%` with the value of the element in `import.meta.env`. So it's the value when executed by JS and not the value written in the source code. From that standpoint, we might say that this is not a breaking change.

> I think overloading `define` to extend `import.meta.env.*` is fine since it's not a common thing to do, but happy to discuss more about it if it's confusing.

To be honest, a few month ago I didn't know that it works like that. I thought it's a simple replacement but if `import.meta.env.*` is defined it affects `import.meta.env` too. (https://github.com/vitejs/vite/issues/12866#issuecomment-1510594056, https://github.com/vitejs/vite/pull/13003)
I think it's confusing that it does more than a simple replacement but I guess it's fine to have this behavior if we add a document about this behavior.

Reply:
> I'm leaning towards applying this suggestion ([#13425 (comment)](https://github.com/vitejs/vite/pull/13425#discussion_r1217804441)).

What happens for `import.meta.env.*` that are numbers or booleans though. I think it makes sense to replace e.g. `<p>%LEGACY%</p>` to `<p>false</p>`

Reply:
I'm wondering whether it's confusing to allow that. But yeah, it would be better to have access to number/boolean `import.meta.env`s.